HeterogeneousCore/AlpakaInterface/interface/prefixScan.h

```diff
@@ -0,0 +1,235 @@
+        auto first = blockThreadIdx;
+        auto laneId = blockThreadIdx & (warpSize - 1);
+
+        for (auto i = first; i < size; i += blockDimension) {
+          warpPrefixScan(acc, laneId, c, i);
+          auto warpId = i / warpSize;
+          ALPAKA_ASSERT_OFFLOAD(warpId < warpSize);
+          if ((warpSize - 1) == laneId)
+            ws[warpId] = c[i];
+        }
+        alpaka::syncBlockThreads(acc);
+        if (size <= warpSize)
+          return;
+        if (blockThreadIdx < warpSize) {
+          warpPrefixScan(acc, laneId, ws, blockThreadIdx);
+        }
+        alpaka::syncBlockThreads(acc);
+        for (auto i = first + warpSize; i < size; i += blockDimension) {
+          auto warpId = i / warpSize;
+          c[i] += ws[warpId - 1];
+        }
+        alpaka::syncBlockThreads(acc);
+      } else {
+        for (int32_t i = 1; i < size; ++i)
+          c[i] += c[i - 1];
+      }
+    }
+
+    // in principle not limited....
+    template <typename T>
+    struct multiBlockPrefixScan {
+      template <typename TAcc>
+      ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                    T const* ci,
+                                    T* co,
+                                    uint32_t size,
+                                    int32_t numBlocks,
+                                    int32_t* pc,
+                                    std::size_t warpSize) const {
+        // Get shared variable. TODO: The workspace is in fact not needed at all for single thread multi element accelerators TODO
+        T* ws = nullptr;
+        if constexpr (!requires_single_thread_per_block_v<TAcc>) {
+          ws = alpaka::getDynSharedMem<T>(acc);
+        }
+        ALPAKA_ASSERT_OFFLOAD(warpSize == static_cast<std::size_t>(alpaka::warp::getSize(acc)));
+        [[maybe_unused]] const auto elementsPerGrid = alpaka::getWorkDiv<alpaka::Grid, alpaka::Elems>(acc)[0u];
+        const auto elementsPerBlock = alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[0u];
+        const auto threadsPerBlock = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
+        const auto blocksPerGrid = alpaka::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc)[0u];
+        const auto blockIdx = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u];
+        const auto threadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u];
+        ALPAKA_ASSERT_OFFLOAD(elementsPerGrid >= size);
+        // first each block does a scan
+        [[maybe_unused]] int off = elementsPerBlock * blockIdx;
+        if (size - off > 0) {
+          blockPrefixScan(acc, ci + off, co + off, std::min(elementsPerBlock, size - off), ws);
+        }
+
+        // count blocks that finished
+        auto& isLastBlockDone = alpaka::declareSharedVar<bool, __COUNTER__>(acc);
+        //__shared__ bool isLastBlockDone;
+        if (0 == threadIdx) {
+          alpaka::mem_fence(acc, alpaka::memory_scope::Device{});
```

Comment:
```suggestion
          alpaka::mem_fence(acc, alpaka::memory_scope::Grid{});
```

`Grid` should be enough, we don't need to have a fence visible to other kernel launches, right ?